# Maji Ndogo Water Services — Part 2: Cleaning & Clustering the Data

In Part 1 we explored the `md_water_services` database. In Part 2 we start **cleaning** it and **clustering** (grouping and summarising) the records to understand how people in Maji Ndogo access water.

**What this notebook covers**

1. [Setup](#1.-Setup:-connect-to-MySQL): connect Python to MySQL
2. [Employee data](#2.-Employee-data): build email addresses, fix phone numbers, find the busiest surveyors
3. [Location data](#3.-Location-data): how records are spread across towns, provinces and rural/urban areas
4. [Water sources](#4.-Water-sources): who uses which source, and which sources to fix first
5. [Queue times](#5.-Queue-times): when people wait longest to collect water
6. [Project 7](#Project-7:-Questions-&-Answers): questions and answers

**Tools:** MySQL 8.0 · Python · SQLAlchemy + PyMySQL · `jupysql` (`%sql` / `%%sql` magics)

---
## 1. Setup: connect to MySQL

Credentials are read from a `.env` file (never hard-coded) with `python-dotenv`. SQLAlchemy builds the connection, and the `sql` extension lets us run queries directly in cells with `%sql` (one line) and `%%sql` (whole cell).

In [1]:
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine
from sqlalchemy.engine import URL

load_dotenv()

url = URL.create(
    drivername="mysql+pymysql",
    username=os.getenv("DB_USER"),
    password=os.getenv("DB_PASSWORD"),
    host=os.getenv("DB_HOST", "localhost"),
    port=3306,
    database=os.getenv("DB_NAME"),
)

engine = create_engine(url)

%load_ext sql
%sql engine

### Check the tables

The database has eight tables: `data_dictionary`, `employee`, `global_water_access`, `location`, `visits`, `water_quality`, `water_source` and `well_pollution`.

In [2]:
%sql show tables

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

8 rows affected.

Tables_in_md_water_services
data_dictionary
employee
global_water_access
location
visits
water_quality
water_source
well_pollution


---
## 2. Employee data

### 2.1 Generate employee emails

Every employee needs an email address in the format `first.last@majindogogov.com`. We build it in four steps: look at the names, transform them, add the domain, then update the table.

**Step 1: look at the names**

In [3]:
%%sql

select employee_name from employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

employee_name
Amara Jengo
Bello Azibo
Bakari Iniko
Malachi Mavuso
Cheche Buhle
Zuriel Matembo
Deka Osumare
Lalitha Kaburi
Enitan Zuri
Farai Nia


**Step 2: make them email-friendly.** `REPLACE` swaps the space for a dot and `LOWER` makes everything lowercase. This is the first draft, so there is no domain yet.

In [4]:
%%sql

select lower(replace(employee_name, ' ', '.'))as email 
from employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

email
amara.jengo
bello.azibo
bakari.iniko
malachi.mavuso
cheche.buhle
zuriel.matembo
deka.osumare
lalitha.kaburi
enitan.zuri
farai.nia


**Step 3: add the domain** with `CONCAT`.

In [5]:
%%sql

select concat(lower(replace(employee_name, ' ', '.')), '@majindogogov.com') as email
from employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

email
amara.jengo@majindogogov.com
bello.azibo@majindogogov.com
bakari.iniko@majindogogov.com
malachi.mavuso@majindogogov.com
cheche.buhle@majindogogov.com
zuriel.matembo@majindogogov.com
deka.osumare@majindogogov.com
lalitha.kaburi@majindogogov.com
enitan.zuri@majindogogov.com
farai.nia@majindogogov.com


**Step 4: save it.** `UPDATE` writes the new addresses into the `email` column of every employee.

In [6]:
%%sql
update employee
set email = concat(lower(replace(employee_name, ' ', '.')), '@majindogogov.com')

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

++
||
++
++

### 2.2 Clean the phone numbers

Phone numbers should be 12 characters long, but the length check shows 13, so there is an extra (invisible) character, most likely a trailing space.

In [7]:
%%sql

select length(phone_number) as phone_length, phone_number from employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

phone_length,phone_number
12,+99637993287
12,+99643864786
12,+99222599041
12,+99945849900
12,+99381679640
12,+99034075111
12,+99379364631
12,+99681623240
12,+99248509202
12,+99570082739


`TRIM` removes leading and trailing spaces. We preview the result first: the length drops from 13 to 12, which confirms the diagnosis.

In [8]:
%%sql

select trim(phone_number) as trimmed_phone_number, length(trim(phone_number)) as trimmed_length from employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

trimmed_phone_number,trimmed_length
+99637993287,12
+99643864786,12
+99222599041,12
+99945849900,12
+99381679640,12
+99034075111,12
+99379364631,12
+99681623240,12
+99248509202,12
+99570082739,12


Now apply the fix permanently.

In [9]:
%%sql

update employee
set phone_number = trim(phone_number)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

++
||
++
++

### 2.3 Where do employees live?

Counting employees per town shows how the workforce is distributed. Most employees (29) are based in `Rural` areas.

In [10]:
%%sql

select town_name, count(*) as count_per_town
from employee
group by town_name


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

9 rows affected.

town_name,count_per_town
Ilanga,3
Rural,29
Lusaka,4
Zanzibar,4
Dahabu,6
Kintampo,1
Harare,5
Yaounde,1
Serowe,3


### 2.4 Top 5 employees by number of visits

Which employees did the most field visits? We count visits per `assigned_employee_id` and keep the top five.

In [12]:
%%sql
select assigned_employee_id, count(*) as count_per_visit
from visits
group by assigned_employee_id
order by count_per_visit desc
limit 5

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

assigned_employee_id,count_per_visit
1,3708
30,3676
34,3539
3,3420
10,3407


Then we pull their contact details so they can be recognised for their work.

In [13]:
%%sql 

select employee_name, phone_number, town_name, email
from employee
where assigned_employee_id in (1, 30, 34, 3, 10)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

employee_name,phone_number,town_name,email
Bello Azibo,+99643864786,Rural,bello.azibo@majindogogov.com
Malachi Mavuso,+99945849900,Lusaka,malachi.mavuso@majindogogov.com
Farai Nia,+99570082739,Dahabu,farai.nia@majindogogov.com
Pili Zola,+99822478933,Zanzibar,pili.zola@majindogogov.com
Rudo Imani,+99046972648,Ilanga,rudo.imani@majindogogov.com


---
## 3. Location data

### 3.1 Records per town and province

In [14]:
%%sql

select town_name, count(*) as records_per_town
from location
group by town_name


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

25 rows affected.

town_name,records_per_town
Harare,1650
Kintampo,780
Lusaka,1070
Rural,23740
Abidjan,400
Amina,1090
Asmara,930
Bello,400
Dahabu,930
Pwani,520


Records per province: five provinces in total (`Akatsi`, `Amanzi`, `Hawassa`, `Kilimani`, `Sokoto`).

In [15]:
%%sql

select province_name, count(*) as records_per_province
from location
group by province_name

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

province_name,records_per_province
Akatsi,8940
Amanzi,6950
Hawassa,6030
Kilimani,9510
Sokoto,8220


Towns sorted from most to fewest records. `Rural` dominates by a wide margin.

In [16]:
%%sql

select town_name, count(*) as records_per_town
from location
group by town_name
order by records_per_town desc

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

25 rows affected.

town_name,records_per_town
Rural,23740
Harare,1650
Amina,1090
Lusaka,1070
Mrembo,990
Asmara,930
Dahabu,930
Kintampo,780
Ilanga,780
Isiqalo,770


Towns within each province: the same counts, broken down by `province_name`.

In [17]:
%%sql

select province_name, town_name, count(*) as records_per_town
from location
group by province_name, town_name
order by province_name, records_per_town desc

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

31 rows affected.

province_name,town_name,records_per_town
Akatsi,Rural,6290
Akatsi,Lusaka,1070
Akatsi,Harare,800
Akatsi,Kintampo,780
Amanzi,Rural,3100
Amanzi,Asmara,930
Amanzi,Dahabu,930
Amanzi,Amina,670
Amanzi,Pwani,520
Amanzi,Abidjan,400


### 3.2 Rural vs urban

In [18]:
%%sql

select location_type, count(*) as number_of_locations
from location
group by location_type

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2 rows affected.

location_type,number_of_locations
Urban,15910
Rural,23740


What share of all locations is rural? About **60%**.

In [50]:
%%sql

select (count(*) / (select count(*) from location)) * 100 as percentage_rural
from location
where location_type = 'rural'

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

percentage_rural
59.8739


---
## 4. Water sources

### 4.1 How many people do we serve in total?

`number_of_people_served` is summed across all sources: roughly **27.6 million** people.

In [20]:
%%sql

select SUM(number_of_people_served) as total_people_served
from water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

total_people_served
27628140


### 4.2 Sources by type

How many of each source type exist? Wells are by far the most common.

In [21]:
%%sql

select type_of_water_source, count(*) as number_of_sources
from water_source
group by type_of_water_source
order by number_of_sources desc

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

type_of_water_source,number_of_sources
well,17383
tap_in_home,7265
tap_in_home_broken,5856
shared_tap,5767
river,3379


On average, how many people use each source type? A `shared_tap` serves about 2,000 people, far more than any other type.

In [22]:
%%sql
select type_of_water_source, round(AVG(number_of_people_served)) as average_people_served
from water_source
group by type_of_water_source
order by type_of_water_source desc

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

type_of_water_source,average_people_served
well,279
tap_in_home_broken,649
tap_in_home,644
shared_tap,2071
river,699


Total people served per source type.

In [23]:
%%sql
select type_of_water_source, sum(number_of_people_served) as total_people_served
from water_source
group by type_of_water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

type_of_water_source,total_people_served
tap_in_home,4678880
tap_in_home_broken,3799720
well,4841724
shared_tap,11945272
river,2362544


The same totals as a **percentage** of the population. Shared taps serve **43%** of everyone, so they are the most important source type in Maji Ndogo.

In [24]:
%%sql

select type_of_water_source, round(sum(number_of_people_served) / (select sum(number_of_people_served) from water_source) * 100) as percentage_people_served_per_type
from water_source
group by type_of_water_source
order by percentage_people_served_per_type desc

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

type_of_water_source,percentage_people_served_per_type
shared_tap,43
well,18
tap_in_home,17
tap_in_home_broken,14
river,9


### 4.3 Ranking sources to fix first

We rank source types by how many people depend on them. `tap_in_home` is excluded because it is already the best-case source, so it does not need fixing.

The next three cells compare three ranking window functions. Here there are no ties between source types, so they all give the same result.

**`RANK()`**: ties share a rank, and the next rank is skipped (1, 1, 3).

In [25]:
%%sql

select 
    type_of_water_source,
    SUM(number_of_people_served) as total_people_served,
    rank () over (order by SUM(number_of_people_served) desc) as priority_rank
from water_source
group by type_of_water_source
having type_of_water_source != 'tap_in_home'
order by priority_rank

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

type_of_water_source,total_people_served,priority_rank
shared_tap,11945272,1
well,4841724,2
tap_in_home_broken,3799720,3
river,2362544,4


**`DENSE_RANK()`**: ties share a rank, but no ranks are skipped (1, 1, 2).

In [26]:
%%sql

select 
    type_of_water_source,
    SUM(number_of_people_served) as total_people_served,
    dense_rank () over (order by SUM(number_of_people_served) desc) as priority_rank
from water_source
group by type_of_water_source
having type_of_water_source != 'tap_in_home'
order by priority_rank

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

type_of_water_source,total_people_served,priority_rank
shared_tap,11945272,1
well,4841724,2
tap_in_home_broken,3799720,3
river,2362544,4


**`ROW_NUMBER()`**: every row gets a unique number, even when values tie.

In [27]:
%%sql

select 
    type_of_water_source,
    SUM(number_of_people_served) as total_people_served,
    row_number () over (order by SUM(number_of_people_served) desc) as priority_rank
from water_source
group by type_of_water_source
having type_of_water_source != 'tap_in_home'
order by priority_rank

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

type_of_water_source,total_people_served,priority_rank
shared_tap,11945272,1
well,4841724,2
tap_in_home_broken,3799720,3
river,2362544,4


### 4.4 Ranking individual sources

The same idea at the level of individual sources (`source_id`). With 32,385 sources, ties are common here, so you can see the difference between ranking functions: `RANK()` gives 1, 1, 3, 3, 3, 6, 6, …

In [28]:
%%sql

select 
    source_id,
    type_of_water_source,
    SUM(number_of_people_served) as total_people_served,
    rank () over (order by SUM(number_of_people_served) desc) as priority_rank
from water_source
group by source_id, type_of_water_source
having type_of_water_source != 'tap_in_home'
order by priority_rank

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

32385 rows affected.

source_id,type_of_water_source,total_people_served,priority_rank
HaRu19509224,shared_tap,3998,1
AkRu05603224,shared_tap,3998,1
AkRu04862224,shared_tap,3996,3
AmAs10911224,shared_tap,3996,3
KiHa22867224,shared_tap,3996,3
KiZu31330224,shared_tap,3994,6
HaRu19839224,shared_tap,3994,6
KiRu28630224,shared_tap,3992,8
KiZu31415224,shared_tap,3992,8
AmPw12653224,shared_tap,3990,10


---
## 5. Queue times

How long do people wait to collect water? We only look at visits where there was a queue (`time_in_queue > 0`).

### 5.1 Average queue time by day

Sorted from longest to shortest wait. **Saturday is by far the longest (~246 min)** and Sunday the shortest (~82 min).

In [48]:
%%sql

select 
    dayname(time_of_record) as day_of_week,
    round(avg(time_in_queue)) as average_time_in_queue
from visits
where time_in_queue > 0
group by day_of_week
order by average_time_in_queue desc


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

7 rows affected.

day_of_week,average_time_in_queue
Saturday,246
Monday,137
Friday,120
Tuesday,108
Thursday,105
Wednesday,97
Sunday,82


### 5.2 Average queue time by hour of day

`HOUR()` extracts the hour from the timestamp. Waits are longest at the start of the day (about 149 min from 06:00 to 08:00) and drop to roughly 111–118 min from 09:00 onwards.

In [49]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(time_in_queue), 0) AS avg_queue_time
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
order by hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,avg_queue_time
6,149
7,149
8,149
9,118
10,114
11,111
12,112
13,115
14,114
15,114


### 5.3 Queue time by hour **and** day: building a pivot table

We want one row per hour and one column per day. The trick is a `CASE WHEN` inside `AVG()` so each column only averages its own day. `ELSE NULL` matters here: `AVG()` ignores `NULL`s, so other days' visits don't pull the average down (`ELSE 0` would count them as zero-minute waits).

We build it one day at a time, then combine all seven. **Sunday:**

In [54]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue ELSE null END), 0) AS Sunday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Sunday
6,79
7,82
8,86
9,84
10,83
11,78
12,78
13,81
14,83
15,83


**Monday:**

In [55]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue ELSE null END), 0) AS Monday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Monday
6,190
7,186
8,183
9,127
10,119
11,115
12,115
13,122
14,127
15,126


**Tuesday:**

In [56]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue ELSE null END), 0) AS Tuesday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Tuesday
6,134
7,128
8,130
9,105
10,99
11,102
12,97
13,97
14,104
15,104


**Wednesday:**

In [57]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue ELSE null END), 0) AS Wednesday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Wednesday
6,112
7,111
8,119
9,94
10,89
11,86
12,88
13,98
14,92
15,88


**Thursday:**

In [58]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue ELSE null END), 0) AS Thursday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Thursday
6,134
7,139
8,129
9,99
10,95
11,99
12,96
13,101
14,96
15,92


**Friday:**

In [59]:
%%sql

SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue ELSE null END), 0) AS Friday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Friday
6,153
7,156
8,153
9,107
10,112
11,104
12,109
13,115
14,110
15,110


**Saturday:**

In [60]:
%%sql

select
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue ELSE null END), 0) AS Saturday
FROM
    visits
WHERE
    time_in_queue > 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Saturday
6,247
7,247
8,247
9,252
10,259
11,236
12,239
13,242
14,244
15,248


**Final version: all seven days in one query.** Each `CASE WHEN` becomes its own column, and `TIME_FORMAT` shows the hour as `06:00`.

In [61]:
%%sql

select time_format(time(time_of_record), '%H:00') as hour_of_day,
    -- sunday
    round(avg(
        case 
        when dayname(time_of_record) = 'Sunday' then time_in_queue 
        else null 
        end)) 
        as Sunday,
 
    -- monday
    round(avg(
        case 
        when dayname(time_of_record) = 'Monday' then time_in_queue 
        else null 
        end)) as Monday,

    -- tuesday
    round(avg(
        case 
        when dayname(time_of_record) = 'Tuesday' then time_in_queue 
        else null 
        end), 0) as Tuesday,

    -- wednesday
    round(avg( 
        case 
        when dayname(time_of_record) = 'Wednesday' then time_in_queue 
        else null 
        end), 0) as Wednesday,

    -- thursday
    round(avg(
        case 
        when dayname(time_of_record) = 'Thursday' then time_in_queue 
        else null 
        end), 0) as Thursday,
        
    -- friday
    round(avg(
        case when dayname(time_of_record) = 'Friday' then time_in_queue 
        else null 
        end), 0) as Friday,

    -- saturday
    round(avg(
        case when dayname(time_of_record) = 'Saturday' then time_in_queue 
        else null 
        end), 0) as Saturday
from visits
where time_in_queue > 0
group by hour_of_day
order by hour_of_day



Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

hour_of_day,Sunday,Monday,Tuesday,Wednesday,Thursday,Friday,Saturday
06:00,79,190,134,112,134,153,247
07:00,82,186,128,111,139,156,247
08:00,86,183,130,119,129,153,247
09:00,84,127,105,94,99,107,252
10:00,83,119,99,89,95,112,259
11:00,78,115,102,86,99,104,236
12:00,78,115,97,88,96,109,239
13:00,81,122,97,98,101,115,242
14:00,83,127,104,92,96,110,244
15:00,83,126,104,88,92,110,248


### Key takeaways: queue times

- **Saturday is the worst day**: waits are roughly 235–260 minutes at every hour from 06:00 to 15:00.
- **Sunday is the best day**: waits stay around 80 minutes all day.
- **Monday to Friday**: the early morning (06:00–08:00) is the busiest. Monday is ~183–190 min at those hours, dropping to ~115–127 min by mid-morning. Midday and afternoon are calmer.
- Only 06:00–15:00 is displayed because the output is truncated to 10 rows; the query covers 14 hours.

---
# Project 7: Questions & Answers

Each question below is followed by the query that answers it and the result.

**Question 1: How many water sources of each type are there?**

*Answer:* `well` 17,383 · `tap_in_home` 7,265 · `tap_in_home_broken` 5,856 · `shared_tap` 5,767 · `river` 3,379.

In [39]:
%%sql

select type_of_water_source, count(*) as user_count
from water_source
group by type_of_water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

type_of_water_source,user_count
tap_in_home,7265
tap_in_home_broken,5856
well,17383
shared_tap,5767
river,3379


**Question 2: How many shared taps are there, and how many people use each one on average?**

*Answer:* There are **5,767** shared taps, each serving about **2,071** people on average.

In [40]:
%%sql

select 
    count(*) as shared_tap_user_count,
    avg(number_of_people_served) as avg_people_served_per_shared_tap
from water_source
where type_of_water_source = 'shared_tap';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

shared_tap_user_count,avg_people_served_per_shared_tap
5767,2071.3147


**Question 3: What percentage of locations are rural?**

*Answer:* About **59.87%**.

In [41]:
%%sql
SELECT
    (COUNT(*) / (SELECT COUNT(*) FROM water_source)) * 100 AS percentage_rural
FROM location
WHERE location_type = 'rural';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

percentage_rural
59.8739


**Question 4: What percentage of locations are urban?**

*Answer:* About **40.13%**.

In [53]:
%%sql
SELECT
    (COUNT(*) / (SELECT COUNT(*) FROM water_source)) * 100 AS percentage_urban
FROM location
WHERE location_type = 'urban';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

percentage_urban
40.1261


**Question 5: For each source type, how many people does it serve in total, and how many per source on average?**

*Answer:* Shared taps serve the most people in total (**11,945,272**, avg 2,071 per tap). Wells serve the fewest per source (avg 279).

In [43]:
%%sql

select 
    type_of_water_source,
    sum(number_of_people_served) as total_people_served,
    avg(number_of_people_served) as average_people_served
from water_source
group by type_of_water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

type_of_water_source,total_people_served,average_people_served
tap_in_home,4678880,644.0303
tap_in_home_broken,3799720,648.8593
well,4841724,278.5321
shared_tap,11945272,2071.3147
river,2362544,699.1844


**Question 6: How many people use each source type (compared with the total population)?**

*Answer:* Total population served: **27,628,140**. Shared tap: 11,945,272 · well: 4,841,724 · tap in home: 4,678,880 · broken tap in home: 3,799,720 · river: 2,362,544.

In [51]:
%%sql

select
    sum(number_of_people_served) as total_population,
    sum(case when type_of_water_source = 'tap_in_home' then number_of_people_served else 0 end) as tap_in_home_population,
    sum(case when type_of_water_source = 'shared_tap' then number_of_people_served else 0 end) as shared_tap_population,
    sum(case when type_of_water_source = 'tap_in_home_broken' then number_of_people_served else 0 end) as tap_in_home_broken_population,
   sum(case when type_of_water_source = 'river' then number_of_people_served else 0 end) as river_population,
    sum(case when type_of_water_source = 'well' then number_of_people_served else 0 end) as well_population
    from water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

total_population,tap_in_home_population,shared_tap_population,tap_in_home_broken_population,river_population,well_population
27628140,4678880,11945272,3799720,2362544,4841724


**Question 7: Show the same population figures in millions.**

*Answer:* Total **27.63M** · shared tap 11.95M · well 4.84M · tap in home 4.68M · broken tap in home 3.80M · river 2.36M.

In [52]:
%%sql
select
    round((sum(number_of_people_served) / 1e6), 2) as total_population_in_millions,
    round((sum(case when type_of_water_source = 'tap_in_home' then number_of_people_served else 0 end) / 1e6), 2) as tap_in_home_population_in_millions,
    round((sum(case when type_of_water_source = 'shared_tap' then number_of_people_served else 0 end) / 1e6), 2) as shared_tap_population_in_millions,
    round((sum(case when type_of_water_source = 'tap_in_home_broken' then  number_of_people_served else 0 end) / 1e6), 2) as tap_in_home_broken_population_in_millions,
    round((sum(case when type_of_water_source = 'river' then number_of_people_served else 0 end) / 1e6), 2) as river_population_in_millions,
    round((sum(case when type_of_water_source = 'well' then number_of_people_served else 0 end) / 1e6), 2) as well_population_in_millions
    from water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

total_population_in_millions,tap_in_home_population_in_millions,shared_tap_population_in_millions,tap_in_home_broken_population_in_millions,river_population_in_millions,well_population_in_millions
27.63,4.68,11.95,3.8,2.36,4.84


**Question 8: What is the average queue time across all visits (ignoring visits with no queue)?**

*Answer:* About **123 minutes**.

In [46]:
%%sql

select
    avg(time_in_queue) as average_time_in_queue
    from visits
    where time_in_queue > 0

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

average_time_in_queue
123.2574


**Question 9: What is the average queue time for each day of the week and hour of the day?**

*Answer:* 98 rows (7 days × 14 hours). Use this to plan visits around the busy periods found in section 5.

In [47]:
%%sql
select 
    dayname(time_of_record) as day_of_week,
    time_format(time(time_of_record), '%H:00') as hour_of_day,
    round(avg(time_in_queue)) as average_time_in_queue
    from visits
    where time_in_queue > 0
    group by day_of_week, hour_of_day
    order by day_of_week, hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

98 rows affected.

day_of_week,hour_of_day,average_time_in_queue
Friday,06:00,153
Friday,07:00,156
Friday,08:00,153
Friday,09:00,107
Friday,10:00,112
Friday,11:00,104
Friday,12:00,109
Friday,13:00,115
Friday,14:00,110
Friday,15:00,110
